relational_operators/UnionAllOperator.hpp

```diff
@@ -0,0 +1,225 @@
+        input_relations_are_stored_(input_relations_are_stored),
+        output_relation_(output_relation),
+        output_destination_index_(output_destination_index),
+        select_attribute_ids_(select_attribute_ids),
+        stored_generated_(false) {
+    // For every input relation, initialize the UnionAll data structure.
+    for (std::size_t i = 0; i < input_relations.size(); i++) {
+      relation_id_to_index_.emplace(input_relations[i]->getID(), i);
+      if (input_relations_are_stored[i]) {
+        input_relations_block_ids_.emplace_back(
+            input_relations[i]->getBlocksSnapshot());
+      } else {
+        input_relations_block_ids_.emplace_back();
+        still_feeding_.insert(i);
+      }
+
+      num_workorders_generated_.push_back(0);
+    }
+  }
+
+  ~UnionAllOperator() override {}
+
+  OperatorType getOperatorType() const override {
+    return kUnionAll;
+  }
+
+  std::string getName() const override {
+    return "UnionAll";
+  }
+
+  QueryContext::insert_destination_id getInsertDestinationID() const override {
+    return output_destination_index_;
+  }
+
+  /**
+   * @return The input relations to this UnionAll operator.
+   */
+  const std::vector<const CatalogRelation*>& input_relations() const {
+    return input_relations_;
+  }
+
+  /**
+   * @return The output relation to this UnionAll operator.
+   */
+  const CatalogRelation& output_relation() const {
+    return output_relation_;
+  }
+
+  const relation_id getOutputRelationID() const override {
+    return output_relation_.getID();
+  }
+
+  void feedInputBlock(const block_id input_block_id,
+                      const relation_id input_relation_id,
+                      const partition_id part_id) override;
+
+  void doneFeedingInputBlocks(const relation_id rel_id) override;
+
+  bool getAllWorkOrders(WorkOrdersContainer *container,
+                        QueryContext *query_context,
+                        StorageManager *storage_manager,
+                        const tmb::client_id scheduler_client_id,
+                        tmb::MessageBus *bus) override;
+
+  bool getAllWorkOrderProtos(WorkOrderProtosContainer *container) override;
+
+ private:
+  // Add work orders for a single relation.
+  void addWorkOrdersSingleRelation(WorkOrdersContainer *container,
+                                   QueryContext *query_context,
+                                   StorageManager *storage_manager,
+                                   InsertDestination *output_destination,
+                                   const std::size_t relation_index);
+
+  const std::vector<const CatalogRelation*> input_relations_;
+  const std::vector<bool> input_relations_are_stored_;
+
+  const CatalogRelation &output_relation_;
+  const QueryContext::insert_destination_id output_destination_index_;
+
+  const std::vector<std::vector<attribute_id>> select_attribute_ids_;
+
+  std::vector<std::vector<block_id>> input_relations_block_ids_;
```

Comment:
Please add some comments stating that the index i represents the blocks from input_relations_[i].